src/zProposal.ts

```diff
@@ -0,0 +1,250 @@
+  private readonly _state: 'pending' | 'active' | 'closed';
+  private readonly _network: string;
+  private readonly _snapshot: string;
+  private readonly _scores: number[];
+  private readonly _votes: number;
+  private _metadata: TokenMetaData | undefined;
+
+  constructor(
+    zDAO: zDAOClient,
+    snapshotClient: SnapshotClient,
+    gnosisSafeClient: GnosisSafeClient,
+    id: ProposalId,
+    type: string,
+    author: string,
+    title: string,
+    body = '',
+    ipfs: string,
+    choices: VoteChoice[],
+    created: Date,
+    start: Date,
+    end: Date,
+    state: 'pending' | 'active' | 'closed',
+    network: string,
+    snapshot: string,
+    scores: number[],
+    votes: number
+  ) {
+    this._zDAO = zDAO;
+    this._snapshotClient = snapshotClient;
+    this._gnosisSafeClient = gnosisSafeClient;
+    this._id = id;
+    this._type = type;
+    this._author = author;
+    this._title = title;
+    this._body = body;
+    this._ipfs = ipfs;
+    this._choices = choices;
+    this._created = created;
+    this._start = start;
+    this._end = end;
+    this._state = state;
+    this._network = network;
+    this._snapshot = snapshot;
+    this._scores = scores;
+    this._votes = votes;
+  }
+
+  get id() {
+    return this._id;
+  }
+
+  get type() {
+    return this._type;
+  }
+
+  get author() {
+    return this._author;
+  }
+
+  get title() {
+    return this._title;
+  }
+
+  get body() {
+    return this._body;
+  }
+
+  get ipfs() {
+    return this._ipfs;
+  }
+
+  get choices() {
+    return this._choices;
+  }
+
+  get created() {
+    return this._created;
+  }
+
+  get start() {
+    return this._start;
+  }
+
+  get end() {
+    return this._end;
+  }
+
+  get state() {
+    return this._state;
+  }
+
+  get network() {
+    return this._network;
+  }
+
+  get snapshot() {
+    return this._snapshot;
+  }
+
+  get scores() {
+    return this._scores;
+  }
+
+  get votes() {
+    return this._votes;
+  }
+
+  get metadata() {
+    return this._metadata;
+  }
+
+  async getTokenMetadata(): Promise<TokenMetaData> {
+    if (!this._ipfs) {
+      throw Error(t('empty-voting-token'));
+    }
+    if (this._metadata) {
+      return this._metadata;
+    }
+
+    const ipfsData = await this._snapshotClient.ipfsGet(this._ipfs);
+    if (!ipfsData.data || !ipfsData.data.message) {
+      throw Error(t('empty-voting-token'));
+    }
+
+    const metadataJson = JSON.parse(ipfsData.data.message.metadata);
+    const sender = metadataJson.sender;
+    const recipient = metadataJson.recipient;
+    const token = metadataJson.token;
+    const decimals = metadataJson.decimals ?? 18;
+    const symbol = metadataJson.symbol ?? 'zToken';
+    const amount = metadataJson.amount;
+
+    this._metadata = {
+      sender,
+      recipient,
+      token,
+      decimals,
+      symbol,
+      amount,
+    };
+    return this._metadata;
+  }
+
+  async listVotes(from = 0, count = 30000, voter = ''): Promise<Vote[]> {
```

Comment:
Some graphql endpoints put limitations on how large `count` can be, and sometimes even ignore it.
Ensure that 30000 is actually the highest count allowed. You may have to create a paginated query. (don't make the user paginate it themselves)